yaml/gateway_endpoint.go

```diff
@@ -0,0 +1,162 @@
+		JWTAuthorizedUsers []string             `yaml:"jwt_authorized_users,omitempty"`
+	}
+	// rateLimitingYAML represents the RateLimiting section of a single GatewayEndpoint in the YAML file.
+	rateLimitingYAML struct {
+		ThroughputLimit     int                             `yaml:"throughput_limit"`
+		CapacityLimit       int                             `yaml:"capacity_limit"`
+		CapacityLimitPeriod grpc_server.CapacityLimitPeriod `yaml:"capacity_limit_period"`
+	}
+)
+
+func (e *gatewayEndpointYAML) convertToProto(endpointID string) *proto.GatewayEndpoint {
+
+	metadata := map[string]string{}
+	for key, value := range e.Metadata {
+		metadata[key] = value
+	}
+
+	return &proto.GatewayEndpoint{
+		EndpointId: endpointID,
+		Auth:       e.Auth.convertToProto(),
+		RateLimiting: &proto.RateLimiting{
+			ThroughputLimit:     int32(e.RateLimiting.ThroughputLimit),
+			CapacityLimit:       int32(e.RateLimiting.CapacityLimit),
+			CapacityLimitPeriod: grpc_server.CapacityLimitPeriods[e.RateLimiting.CapacityLimitPeriod],
+		},
+		Metadata: metadata,
+	}
+}
+
+func (a *authYAML) convertToProto() *proto.Auth {
+	authProto := &proto.Auth{
+		AuthType: proto.Auth_AuthType(proto.Auth_AuthType_value[string(a.AuthType)]),
+	}
+
+	switch a.AuthType {
+
+	case grpc_server.AuthTypeAPIKey:
+		if a.APIKey != nil {
+			authProto.AuthTypeDetails = &proto.Auth_ApiKey{
+				ApiKey: *a.APIKey,
+			}
+		}
+
+	case grpc_server.AuthTypeJWT:
+		if a.JWTAuthorizedUsers != nil {
+			jwtDetails := &proto.Auth_Jwt{
+				Jwt: &proto.JWT{AuthorizedUsers: make(map[string]*proto.Empty)},
+			}
+			for _, user := range a.JWTAuthorizedUsers {
+				jwtDetails.Jwt.AuthorizedUsers[user] = &proto.Empty{}
+			}
+			authProto.AuthTypeDetails = jwtDetails
+		}
+
+	default:
+		authProto.AuthTypeDetails = &proto.Auth_NoAuth{}
+
+	}
+
+	return authProto
+}
+
+// gatewayEndpoint.validate ensures all fields set for the gateway endpoint are valid.
+func (e *gatewayEndpointYAML) validate(endpointID string) error {
+	if endpointID == "" {
+		return fmt.Errorf("endpoint_id is required")
+	}
+	if err := e.Auth.validate(); err != nil {
+		return err
+	}
+	if err := e.RateLimiting.validate(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// authYAML.validate ensures that the auth section of a GatewayEndpoint is valid by
+// checking that the correct fields are set for the given auth type and are not set
+// for any other auth type.
+func (a *authYAML) validate() error {
+	switch a.AuthType {
+
+	// API Key authorization requires an API key to be set for the endpoint.
+	case grpc_server.AuthTypeAPIKey:
+		if len(a.JWTAuthorizedUsers) > 0 {
+			return fmt.Errorf("jwt_authorized_users must not be set for auth_type: API_KEY_AUTH")
+		}
+		if a.APIKey == nil || *a.APIKey == "" {
+			return fmt.Errorf("api_key is required for auth_type: API_KEY_AUTH")
+		}
+
+	// JWT authorization requires a list of authorized user IDs to be set for the endpoint.
+	case grpc_server.AuthTypeJWT:
```

Comment:
Should we add a base64 decoding check? Feel free to add a TODO

Reply:
I don't follow; what is the purpose of this check in this context?

Reply:
![Screenshot 2024-11-27 at 1 58 24 PM](https://github.com/user-attachments/assets/bba817d5-ceda-4a73-ada7-69d73260d516)

Ensuring that it's a valid JWT?

Reply:
This part doesn't have anything to do with the JWT itself. The JWT is never stored on the GatewayEndpoint struct; it's just attached to the request header and that handled by the PATH Auth Server.